src/OpenApi/src/Services/Schemas/OpenApiSchemaStore.cs

```diff
@@ -35,6 +60,122 @@ internal sealed class OpenApiSchemaStore
+        if (schema.Properties is not null)
+        {
+            foreach (var property in schema.Properties.Values)
+            {
+                AddOrUpdateSchemaByReference(property);
+            }
+        }
+    }
+
+    private void AddOrUpdateSchemaByReference(OpenApiSchema schema)
+    {
+        if (_schemasWithReference.TryGetValue(schema, out var referenceId))
+        {
+            // If we've already used this reference ID else where in the document, increment a counter value to the reference
+            // ID to avoid name collisions. These collisions are most likely to occur when the same .NET type produces a different
+            // schema in the OpenAPI document because of special annotations provided on it. For example, in the two type definitions
+            // below:
+            // public class Todo
+            // {
+            //     public int Id { get; set; }
+            //     public string Name { get; set; }
+            // }
+            // public class Project
+            // {
+            //     public int Id { get; set; }
+            //     [MinLength(5)]
+            //     public string Title { get; set; }
+            // }
+            // The `Title` and `Name` properties are both strings but the `Title` property has a `minLength` annotation
+            // on it that will materialize into a different schema.
+            // {
+            //
+            //      "type": "string",
+            //      "minLength": 5
+            // }
+            // {
+            //      "type": "string"
+            // }
+            // In this case, although the reference ID  based on the .NET type we would use is `string`, the
+            // two schemas are distinct.
+            if (referenceId == null)
+            {
+                var targetReferenceId = GetSchemaReferenceId(schema);
+                if (_referenceIdCounter.TryGetValue(targetReferenceId, out var counter))
+                {
+                    counter++;
+                    _referenceIdCounter[targetReferenceId] = counter;
+                    _schemasWithReference[schema] = $"{targetReferenceId}{counter}";
+                }
+                else
+                {
+                    _referenceIdCounter[targetReferenceId] = 1;
+                    _schemasWithReference[schema] = targetReferenceId;
+                }
+            }
+        }
+        else
+        {
+            _schemasWithReference[schema] = null;
+        }
+    }
+
+    private static string GetSchemaReferenceId(OpenApiSchema schema)
+    {
+        if (schema.Extensions.TryGetValue(OpenApiConstants.SchemaId, out var referenceIdAny)
+            && referenceIdAny is OpenApiString { Value: string referenceId })
+        {
+            return referenceId;
+        }
+
+        throw new InvalidOperationException("The schema reference ID must be set on the schema.");
+    }
+
+    public Dictionary<OpenApiSchema, string?> SchemasByReference => _schemasWithReference;
```

Comment:
Should this be an ImmutableDictionary? Also, should we just make it an internal property instead of a field+property?

Reply:
Simplified this to a property. I'm not the biggest fan of `ImmutableDictionary` given the requirements on using the builder pattern for it, especially because we're building up the list of schemas as the schema service gets called.